embassy-microchip/src/i2c.rs

```diff
@@ -0,0 +1,1183 @@
+        }
+
+        Ok(())
+    }
+
+    async fn write_async_internal(&mut self, address: u8, write: &[u8], send_stop: bool) -> Result<(), Error> {
+        if write.is_empty() {
+            return Err(Error::InvalidWriteBufferLength);
+        }
+
+        self.start_async(address, false, false).await?;
+
+        for byte in write.iter() {
+            self.write_byte_async(*byte).await?;
+        }
+
+        if send_stop {
+            Self::stop();
+        }
+
+        Ok(())
+    }
+
+    /// Read from address into read asynchronously.
+    pub async fn read_async(&mut self, address: u8, read: &mut [u8]) -> Result<(), Error> {
+        self.read_async_internal(address, read, false, true).await
+    }
+
+    /// Write to address from write asynchronously.
+    pub async fn write_async(&mut self, address: u8, write: &[u8]) -> Result<(), Error> {
+        self.write_async_internal(address, write, true).await
+    }
+
+    /// Write to address from write and read from address into read asynchronously.
+    pub async fn write_read_async(&mut self, address: u8, write: &[u8], read: &mut [u8]) -> Result<(), Error> {
+        self.write_async_internal(address, write, false).await?;
+        self.read_async_internal(address, read, true, true).await
+    }
+}
+
+/// Interrupt handler.
+pub struct InterruptHandler<T: Instance> {
+    _phantom: PhantomData<T>,
+}
+
+impl<T: Instance> interrupt::typelevel::Handler<T::Interrupt> for InterruptHandler<T> {
+    unsafe fn on_interrupt() {
+        pac::ECIA.src13().write_value(1 << T::irq_bit());
+
+        T::regs().cfg().modify(|w| {
+            w.set_enmi(false);
+            w.set_enidi(false);
+            w.set_en_aas(false);
+        });
+
+        T::waker().wake();
+    }
+}
+
+impl<'d, T: Instance + 'd, M: Mode> I2c<'d, T, M> {
+    fn new_inner(_peri: Peri<'d, T>, _scl: Peri<'d, AnyPin>, _sda: Peri<'d, AnyPin>, config: Config, port: u8) -> Self {
+        let mut i2c = Self {
+            _peri,
+            _scl,
+            _sda,
+            config,
+            port,
+            phantom: PhantomData,
+        };
+        i2c.reset_reconfigure();
+        i2c
+    }
+
+    fn reset_reconfigure(&mut self) {
+        let r = T::regs();
+
+        let addr1 = self.config.addr1.unwrap_or(0);
+        let addr2 = self.config.addr2.unwrap_or(0);
+
+        critical_section::with(|_| {
+            // Reset the controller first.
+            T::reset();
+
+            r.cfg().write(|w| {
+                w.set_flush_sxbuf(true);
+                w.set_flush_srbuf(true);
+                w.set_flush_mxbuf(true);
+                w.set_flush_mrbuf(true);
+            });
+
+            r.wctrl().write(|w| w.set_pin(true));
+            r.own_addr().write(|w| {
+                w.set_addr1(addr1);
+                w.set_addr2(addr2);
+            });
+
+            r.cfg().write(|w| {
+                if addr1 == 0 || addr2 == 0 {
+                    w.set_gc_dis(false);
```

Comment:
Only if `config.addr1 == Some(0) || config.addr2 == Some(0)`?

Reply:
Yeah, this is an interesting part of microchip's i2c controller. The thing is always acting as both slave and master. When working as a slave, it can respond to two different addresses. If any of those addresses is 0, then it can't respond to a General Call.

I could default to `None` and disable General Call by default, but that means `own_addr` above will write `0` to both addresses.